tests/python/relax/test_inline_functions.py

```diff
@@ -0,0 +1,363 @@
+        def main(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            with R.dataflow():
+                B = A * A
+                C = Before.subroutine(B)
+                D = C + C
+                R.output(D)
+            return D
+
+        @R.function(private=True)
+        def subroutine(B: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            with R.dataflow():
+                C = R.concat([B, B], axis=1)
+                R.output(C)
+            return C
+
+    @R.function(private=True)
+    def expected(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+        with R.dataflow():
+            B = A * A
+            C = R.concat([B, B], axis=1)
+            D = C + C
+            R.output(D)
+        return D
+
+    after = Before["main"].inline_functions({"subroutine": Before["subroutine"]})
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_inline_non_dataflow_block_into_dataflow_block():
+    """Function inlining may not produce invalid Relax IR
+
+    A subroutine call may appear within a DataflowBlock, even if the
+    subroutine does not itself use a DataflowBlock.  In this case, to
+    avoid inserting a non-dataflow block in the middle of a set of
+    dataflow bindings, the DataflowBlock in the caller must be split
+    up.
+    """
+
+    @I.ir_module
+    class Before:
+        @R.function(private=True)
+        def main(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            with R.dataflow():
+                B = A * A
+                C = Before.subroutine(B)
+                D = C + C
+                R.output(D)
+            return D
+
+        @R.function(private=True)
+        def subroutine(B: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+            C = R.concat([B, B], axis=1)
+            return C
+
+    @R.function(private=True)
+    def expected(A: R.Tensor([16, 16], "int32")) -> R.Tensor([16, 32], "int32"):
+        # DataflowBlock before subroutine
+        with R.dataflow():
+            B = A * A
+            R.output(B)
+
+        # BindingBlock from the inlined subroutine.  Because B is used
+        # here, outside of a DataflowBlock, this requires it to be
+        # updated from a DataflowVar to a normal Var.
+        C = R.concat([B, B], axis=1)
+
+        # Resuming the DataflowBlock after the inlined subroutine
+        with R.dataflow():
+            D = C + C
+            R.output(D)
+        return D
+
+    after = Before["main"].inline_functions({"subroutine": Before["subroutine"]})
+    tvm.ir.assert_structural_equal(expected, after)
+
+
+def test_subroutine_with_symbolic_vars():
+    """Inlined subroutines should use the caller's symbolic variables
+
+    Before inlining, the subroutine and the caller have distinct
+    `tir::Var` for each symbolic variables.  After inlining, only the
+    caller's `tir::Var` symbolic variables should remain.
+    """
```

Comment:
What logic in the pass handles this replacement? The use of CanonicalizeBindings per "option 4"?

Reply:
Ah, very nice.